docs/resources/rules/assertions.md

```diff
@@ -0,0 +1,487 @@
+        message: x-codeSamples and x-internal must not be defined
+        severity: error
+        disallowed:
+          - x-code-samples
+          - x-internal
+```
+
+### `defined` example
+
+The following example asserts that `x-codeSamples` is defined.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Operation
+        property: x-codeSamples
+        message: x-codeSamples must be defined
+        severity: error
+        defined: true
+```
+
+### `undefined` example
+
+The following example asserts that `x-code-samples` is undefined.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Operation
+        property: x-code-samples
+        message: x-code-samples is deprecated
+        suggest: 
+          - x-codeSamples instead of x-code-samples
+        severity: error
+        undefined: true
+```
+
+### `nonEmpty` example
+
+The following example asserts that the operation summary is not empty.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Operation
+        property: summary
+        message: Operation summary should not be empty
+        severity: error
+        nonEmpty: true
+```
+
+### `minLength` example
+
+The following example asserts that the minimum length of each operation summary is 20 characters.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Operation
+        property: summary
+        message: Operation summary must have minimum of 20 chars length
+        severity: error
+        minLength: 20
+```
+
+### `maxLength` example
+
+The following example asserts that the maximum length of each operation summary is 20 characters.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Operation
+        property: summary
+        message: Operation summary must have a maximum of 20 characters
+        severity: error
+        maxLength: 20
+```
+
+### `sortOrder` example
+
+Sort order can be applied on lists (arrays) of strings or lists of objects (collections).
+
+The following example asserts that the status schema's enum list is in alphabetical order.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Schema
+        property: enum
+        context:
+          - type: Schema
+            filterKeys: 
+              - status
+        message: The status enums should be in alphabetical order
+        suggest:
+          - alphabetize the list of status enums
+        severity: error
+        sortOrder: asc
+```
+
+The following example asserts the tag names (tags are a collection) are sorted in alphabetical order.
+
+The `property` field is **REQUIRED** to sort collections based on that property's value. 
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - subject: Tag
+        property: name
+        message: Tags should be ordered by name in alphabetical order
+        severity: error
+        sortOrder:
+          direction: asc
+          property: name
+```
+<!-- TODO: Does the sortOrder need a property "property" now? -->
```

Comment:
Yep, you're right, now a property "property" is redundant here; 
and we can use the simpler `sortOrder` notation like `sortOrder: asc`, because we can set the order's property right after the `subject`.

Reply:
I've adjusted it (I think)

Reply:
We need to look at other options. I'm pinging @RomanHotsiy (and discussed it with him in a meeting).

Reply:
I don't like the extra `property` in the `sortOrder`. This will be confusing. We should come up with some better idea.

I think we are mixing a few things here. Sorting by a property value and sorting the property value itself.

I don't think we should rely on the type of the property value to determine the behaviour because some properties can have mixed values, e.g. `type` can be string or array.

One possible solution is to bind the behaviour to our type tree. We clearly know whether some type is a collection or a map (usually these have `_List` or `_Map` suffix). So if the type is a collection we should use `property` as a sort key vs sort the `property` itself:

**Sort collection by property `name`:**
```yaml
lint:
  rules:
    assertions:
      - subject: Tag_List
        property: name
        sortOrder: asc
```

**Sort value of a property:**
```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: enum
        sortOrder: asc
```

The main concern with this approach is we can't actually sort something that is untyped (e.g. vendor extensions). Imagine x-metadata vendor extension which may look like:

```yaml
x-metadata:
  - name: Description
    value: Lorem ipsum
  - name: Location
    value: Lviv
```

```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: x-metadata # how to sort by `name` ?
        sortOrder: asc # what sort order even means here ?
```

Reply:
Here is another idea. What about two different sort assertions:

- `sortCollection`
- `sortValues`

### Sorts the **collection** by the property `name`
```yaml
lint:
  rules:
    assertions:
      - subject: Tag
        property: name
        sortCollection: asc
```

### Sorts the **values** of the property `enum`
```yaml
lint:
  rules:
    assertions:
      - subject: Schema
        property: enum
        sortValues: asc
```

For something that may have mixed values, like `product` in this example:

```yaml
x-metadata:
  - name: Sample name as a string
     product: elf
  - name: What is this?
     product:
        - bar
        - foo
```

Someone can specify if they want to sort by the **values** or by the collection by choosing the appropriate assertion `sortValues` or `sortCollection`. 

The open question is what is the logic of `sortCollection` when the value is not a scalar. In the example above, will it use the array's first value for the comparison logic? Or what?

Reply:
> One possible solution is to bind the behaviour to our type tree. We clearly know whether some type is a collection or a map (usually these have _List or _Map suffix). So if the type is a collection we should use property as a sort key vs sort the property itself:

One thing I don't like about this idea is that we already have types for some maps like `ResponsesMap`, so this feels like we missed types like `TagsMap`. I think we are missing content (tech docs) on the type tree. And this idea may also make things like typeahead less intuitive too.

Reply:
> What about two different sort assertions:

I don't like this idea because `sortValues` and `sortCollection` feel more like a decorator, not as an assertion.

> The open question is what is the logic of sortCollection when the value is not a scalar. 

Also `sortCollection` has other open questions. What if the subject is not part of any collection.

> so this feels like we missed types like TagsMap

What is `TagsMap`? Tags are not a map, it's an array (list). There is `Tag_List`.